controllers/util/k8sutil.go

```diff
@@ -289,3 +292,51 @@ func NewK8sClient(scheme *runtime.Scheme) (client.Client, error) {
 
 	return k8sClient, nil
 }
+
+func CreateOrReplace(ctx context.Context, c client.Client, obj client.Object, f controllerutil.MutateFn) error {
+	key := client.ObjectKeyFromObject(obj)
+	if err := c.Get(ctx, key, obj); err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+		if err := mutate(f, key, obj); err != nil {
+			return err
+		}
+		if err := c.Create(ctx, obj); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	existing := obj.DeepCopyObject()
+	if err := mutate(f, key, obj); err != nil {
+		return err
+	}
+
+	if equality.Semantic.DeepEqual(existing, obj) {
```

Comment:
Will not this always return false as we are also checking the metadata and objectmeta which will differ?

Reply:
We are applying the mutate function after getting the current object. This is make sure that the fields we want are preserved and additional labels/annotations added by users are not removed